Text disappearance

Soreine opened the pull request:
Fix #5 

- [x] Track time and subscribe to browser's animation frames.
- [x] Timestamp inputs so we can style them accordingly
- [x] Implement fading effect 
  - [x] Basic
  - [x] Gorgeous
- [x] Apply it
  - [x] on a line basis first
  - [x] on a character basis
- [x] Cleanup objects once they have totally disappeared
  - [x] Remove them from model
  - [x] Keep the transition smooth (we don't want the input area to make jumps as lines get removed)

Soreine committed:
Track time on AnimationFrame

Soreine committed:
Fading lines

Soreine committed:
Add util `ratio`

Soreine commented:
Deploying prototype to gh-pages now.

Soreine commented:
In order to have smooth animation when lines get removed, a clean solution is to have the bottom of the paragraph fixed in position (so instead of positioning absolutely the `top` and `left`, we could position `left` and `bottom`).
By the way... this makes the assumption that the user is writing left-to-right and top-to-bottom.
Someday we will have to think about internationalising it, to support Chinese and Arabic

mpizenberg commented:
What about supporting handwriting ^^ ?

mpizenberg commented:
I think [this presentation](https://www.youtube.com/watch?v=DsDwYqsLU3E) from last elm conf is worth mentionning here. It's [Matthew Griffith](https://github.com/mdgriffith)'s work (a Matthew so it gets to be awesome ^^).
 - [elm-style-animation](https://github.com/mdgriffith/elm-style-animation)
 - [elm-animation-pack](https://github.com/mdgriffith/elm-animation-pack)

mpizenberg commented on main.elm:
```diff
@@ -30,15 +30,23 @@ type alias Coord =
 
+{-| Dated values
+-}
+type alias Dated a =
```
When possible, I avoid records and replace them by data types. Here, it could be
```elm
type Dated a = Dated Time a
```
Dated also serves as a type constructor then:
```elm
line = Dated someDate someString
```

Soreine replied:
Yeah I could not find the doc about that, nor an example of it. I was not sure it was doable... But that's what I wanted to do :3

Soreine replied:
What's left to extract the value then? Except pattern matching?

Soreine replied:
I would not know how to define myself a type like `List a` for example...

mpizenberg replied:
When a type is defined with only one constructor like my `Date` example, you can pattern match on it directly in function definition like:
```elm
fadingText : Time -> Dated String -> Html Msg
fadingText now (Dated time str) =
    ...
```
Small guide to destructuring (pattern matching) in elm: https://gist.github.com/yang-wei/4f563fbf81ff843e8b1e

Soreine replied:
Good to know that we can destructure using `let`...

Soreine commented:
How do you want to support handwriting? Not everyone has access to a graphic tablet

Soreine committed:
Change Dated to not a Record

Soreine committed:
Remove custom location logic

It will be restored for multi-user experience #8.

Soreine committed:
Sort imports

Soreine committed:
Add debugging for focus fails

Soreine committed:
Per character fading... Not good.

Also prevents backspace etc. Dead keys not working anymore...

Soreine commented:
Deploying... I'm not happy with the current implementation.

Bugs:
- Newlines are removing the last characters.
- The last character never disappear, which is weird.
- Dead keys (accents on QWERTY for example) no longer work.

Good:
- Characters fade individually
- The text is now centered

Bad:
- I disallowed backspace etc. You cannot modify what has been written. And it's so frustrating and not natural. Because typos happen you know, and it's instinctive to correct them... I think we should allow backspaces etc, but just for the current line or something...

Soreine commented:
Concerning the implementation, I think it would be best to have one function that maps the current strokes (timestamped characters etc.) in the model to the hidden input value. And another function that updates the current strokes from the new input value.
The characters that are appended are detected and added to the list.
The characters that are changed are "redrawn", so their timestamp is updated.

We should move out from the logic I made, where the input value is consumed into the model, and the model never affect the current input value.

(Edit: I'm on it)

Soreine commented:
While I finish working on that, if someone finds a better effect than a linear progression of opacity, it would be great :) 
At least something non linear would better match what is in the movie.

For something more advanced, maybe something with a canvas + text clipping, to make awesome effects within the shapes of characters?

Soreine committed:
Refactor input logic

We now have a two-way mapping between what is in the input textarea and
the internal model which is a list of "strokes".

Soreine committed:
Disallow a few movement keys

Soreine committed:
Disable text selection

Soreine commented:
Deployed.

The code is cleaner now.

Soreine committed:
Cleanup old lines

Soreine commented:
Some inspirations...
http://freefrontend.com/css-text-effects/#animated-text

Soreine committed:
Add blur

Soreine commented:
@mpizenberg ready to be reviewed and merged

Soreine committed:
Simpler logError

Soreine committed:
Larger blur

Soreine committed:
Make cursor disappear during typing

Soreine commented on main.elm:
```diff
@@ -156,21 +156,11 @@ focusInput =
 
 -- SUBSCRIPTIONS
+logError =
+    Result.mapError (\err -> Debug.log (toString err) err)
```
I could not do better than that @mpizenberg I'm sorry. I'm bad

mpizenberg commented:
wow, you were motivated! (I'll try to look at the code this weekend, promess ^^)
